components/layout_2020/flow/mod.rs

```diff
@@ -1414,14 +1411,15 @@ impl PlacementState {
     ) {
         self.place_fragment(fragment, sequential_layout_state);
 
-        match fragment {
-            Fragment::Box(box_fragment) => {
-                if let Some(baseline) = box_fragment.last_baseline_offset {
-                    self.last_inflow_baseline_offset =
-                        Some(baseline + box_fragment.content_rect.start_corner.block);
-                }
-            },
-            _ => {},
+        if let Fragment::Box(box_fragment) = fragment {
+            let box_block_offset = box_fragment.content_rect.start_corner.block.into();
+            match (self.inflow_baselines.first, box_fragment.baselines.first) {
+                (None, Some(first)) => self.inflow_baselines.first = Some(first + box_block_offset),
+                _ => {},
+            }
+            if let Some(last) = box_fragment.baselines.last {
+                self.inflow_baselines.last = Some(last + box_block_offset);
+            }
```

Comment:
Possibly clearer like this?

```suggestion
            let Baselines { first, last } = box_fragment.baselines;
            if self.inflow_baselines.first.is_none() {
                self.inflow_baselines.first = first.map(|first| first + box_block_offset);
            }
            self.inflow_baselines.last = last.map(|last| last + box_block_offset);
```

Reply:
Hrm. The problem is that we do not want to set last to `None` if last is already set.